operator/cilium_node.go

```diff
@@ -20,16 +26,240 @@ import (
+func (cfgBuilder *ciliumNodeWatcherCfgBuilder) WithNodeManager(nodeManager allocator.NodeEventHandler) *ciliumNodeWatcherCfgBuilder {
+	// We shouldn't change modify cfgBuilder.nodeManager after a
+	// ciliumNodeWatcherCfg was created with this ciliumNodeWatcherCfgBuilder
+	// because that field will be used to handle Kubernetes events.
+	if cfgBuilder.cfgBuilt.Load().(bool) {
+		panic("node manager already set")
+	}
+	cfgBuilder.nodeManager = nodeManager
+	return cfgBuilder
+}
+
+// WithKVStoreSync sets the ciliumNodeWatcherCfgBuilder to be configured
+// with the KVStoreSync support.
+func (cfgBuilder *ciliumNodeWatcherCfgBuilder) WithKVStoreSync() *ciliumNodeWatcherCfgBuilder {
+	cfgBuilder.kvStoreSync = true
+	return cfgBuilder
+}
+
+// Build builds and if WithKVStoreSync was called it will start a SharedStore.
+// If WithKVStoreSync was called, this function must be executed after setting
+// up a KVStore.
+func (cfgBuilder *ciliumNodeWatcherCfgBuilder) Build() (cfg *ciliumNodeWatcherCfg, err error) {
+	defer func() {
+		if err == nil {
+			cfgBuilder.cfgBuilt.Store(true)
+		}
+	}()
+	cfg = &ciliumNodeWatcherCfg{
+		resourceEventHandler:  cache.ResourceEventHandlerFuncs{},
+		ciliumNodeConvertFunc: k8s.ConvertToCiliumNode,
+		withKVStore:           cfgBuilder.kvStoreSync,
+	}
+
+	// KVStore is enabled -> we will run the event handler to sync objects into
+	// KVStore.
+	if cfgBuilder.kvStoreSync {
+		ciliumNodeKVStore, err := store.JoinSharedStore(store.Configuration{
+			Prefix:     nodeStore.NodeStorePrefix,
+			KeyCreator: nodeStore.KeyCreator,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// Start a go routine to GC all CiliumNodes from the KVStore that are
+		// no longer running.
+		go func() {
+			<-k8sCiliumNodesCacheSynced
+			// Since we processed all events received from k8s we know that
+			// at this point the list in ciliumNodeStore should be the source of
+			// truth and we need to delete all nodes in the kvNodeStore that are
+			// *not* present in the ciliumNodeStore.
+			listOfCiliumNodes := ciliumNodeStore.ListKeys()
+
+			kvStoreNodes := ciliumNodeKVStore.SharedKeysMap()
+			for _, ciliumNode := range listOfCiliumNodes {
+				// The remaining kvStoreNodes are leftovers that need to be GCed
+				kvStoreNodeName := nodeTypes.GetKeyNodeName(option.Config.ClusterName, ciliumNode)
+				delete(kvStoreNodes, kvStoreNodeName)
+			}
+
+			for _, kvStoreNode := range kvStoreNodes {
+				// Only delete the nodes that belong to our cluster
+				if strings.HasPrefix(kvStoreNode.GetKeyName(), option.Config.ClusterName) {
+					ciliumNodeKVStore.DeleteLocalKey(context.TODO(), kvStoreNode)
+				}
+			}
+		}()
+
+		// nodeManager is nil so we don't need to handle any events managed by
+		// the nodeManager and the events from Kubernetes will only be relevant
+		// for KVStore.
+		if cfgBuilder.nodeManager == nil {
+			cfg.resourceEventHandler = cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
```

Comment:
I think for readability take, pull out these function literals to somewhere outside the if/else block and give them semantic meaning. Such as "AddFuncNilNodeManager"